File: backend/app/services/task_service.py
from __future__ import annotations

from fastapi import HTTPException, status

from app.models import ExecutionRecord, Task, TaskCreate, TaskEnabledUpdate, TaskUpdate
from app.services.codex_service import CodexBridgeError, CodexService
from app.services.schedule_utils import get_next_run_at, now_in_app_timezone, validate_schedule
from app.services.task_repository import TaskRepository


class TaskService:
    def __init__(self, repository: TaskRepository, codex_service: CodexService) -> None:
        self.repository = repository
        self.codex_service = codex_service

    def list_tasks(self) -> list[Task]:
        return self.repository.list_tasks()

    def get_task(self, task_id: str) -> Task:
        task = self.repository.get_task(task_id)
        if not task:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found.")
        return task

    def create_task(self, payload: TaskCreate) -> Task:
        try:
            validate_schedule(payload.schedule)
        except ValueError as error:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(error)) from error

        try:
            session_id = self.codex_service.create_session()
        except CodexBridgeError as error:
            raise HTTPException(
                status_code=status.HTTP_502_BAD_GATEWAY,
                detail=f"Failed to create Codex thread: {error}",
            ) from error

        now = now_in_app_timezone()
        next_run_at = get_next_run_at(payload.schedule, now)
        return self.repository.create_task(
            schedule=payload.schedule,
            session_id=session_id,
            prompt=payload.prompt,
            enabled=True,
            created_at=now,
            next_run_at=next_run_at,
        )

    def update_task(self, task_id: str, payload: TaskUpdate) -> Task:
        task = self.get_task(task_id)
        try:
            validate_schedule(payload.schedule)
        except ValueError as error:
            raise HTTPException(status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(error)) from error

        now = now_in_app_timezone()
        next_run_at = get_next_run_at(payload.schedule, now) if task.enabled else None
        updated = self.repository.update_task(
            task_id=task.id,
            schedule=payload.schedule,
            prompt=payload.prompt,
            updated_at=now,
            next_run_at=next_run_at,
        )
        return updated or self.get_task(task_id)

    def set_enabled(self, task_id: str, payload: TaskEnabledUpdate) -> Task:
        task = self.get_task(task_id)
        now = now_in_app_timezone()
        next_run_at = get_next_run_at(task.schedule, now) if payload.enabled else None
        updated = self.repository.set_enabled(task_id, payload.enabled, now, next_run_at)
        return updated or self.get_task(task_id)

    def delete_task(self, task_id: str) -> None:
        self.get_task(task_id)
        self.repository.delete_task(task_id)

    def list_execution_history(self, task_id: str | None = None) -> list[ExecutionRecord]:
        if task_id:
            self.get_task(task_id)
        return self.repository.list_execution_history(task_id=task_id)

File: backend/src/task-service.ts
import crypto from "node:crypto";
import path from "node:path";

import { DEFAULT_WORKSPACE_DIRECTORY } from "./config.js";
import type { CodexHistoryService } from "./codex-history-service.js";
import { AppError, CodexServiceError } from "./errors.js";
import { getNextRunAt, nowInAppTimezone, validateSchedule } from "./schedule-utils.js";
import type { CodexService } from "./codex-service.js";
import type { SkillService } from "./skill-service.js";
import type { TaskRepository } from "./task-repository.js";
import type { TaskChatPayload, TaskEnabledPayload, TaskPayload, TaskSettingsPayload } from "./types.js";

function normalizeEnvironmentVariables(value: Record<string, string> | undefined): Record<string, string> {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return {};
  }

  return Object.fromEntries(
    Object.entries(value)
      .filter((entry): entry is [string, string] => typeof entry[0] === "string" && typeof entry[1] === "string")
      .map(([key, entryValue]) => [key.trim(), entryValue])
      .filter(([key]) => key.length > 0)
  );
}

function buildTaskWorkspaceDirectory(taskId: string): string {
  return path.join(DEFAULT_WORKSPACE_DIRECTORY, taskId);
}

export class TaskService {
  taskRepository: TaskRepository;
  codexService: CodexService;
  skillService: SkillService;
  codexHistoryService: CodexHistoryService | null;

  constructor(
    taskRepository: TaskRepository,
    codexService: CodexService,
    skillService: SkillService,
    codexHistoryService: CodexHistoryService | null
  ) {
    this.taskRepository = taskRepository;
    this.codexService = codexService;
    this.skillService = skillService;
    this.codexHistoryService = codexHistoryService;
  }

  listTasks() {
    return this.taskRepository.listTasks();
  }

  getTask(taskId: string) {
    const task = this.taskRepository.getTask(taskId);
    if (!task) {
      throw new AppError(404, "Task not found.");
    }
    return task;
  }

  async createTask(payload: TaskPayload) {
    validateTaskPayload(payload);
    const taskId = crypto.randomUUID();
    const workspaceDirectory = buildTaskWorkspaceDirectory(taskId);
    const environmentVariables = normalizeEnvironmentVariables(payload.environment_variables);

    try {
      this.skillService.ensureTaskSettingsSkillInstalled(workspaceDirectory);
    } catch (error) {
      throw new AppError(
        500,
        `Failed to install default task skill: ${error instanceof Error ? error.message : String(error)}`
      );
    }

    let threadId: string;
    try {
      threadId = await this.codexService.createThread(workspaceDirectory, environmentVariables);
    } catch (error) {
      const message = error instanceof CodexServiceError ? error.message : String(error);
      throw new AppError(502, `Failed to create Codex thread: ${message}`);
    }

    return this.createTaskRecord(taskId, threadId, workspaceDirectory, environmentVariables, payload);
  }

  updateTask(taskId: string, payload: TaskPayload) {
    return this.updateTaskSettings(taskId, payload);
  }

  updateTaskSettings(taskId: string, payload: TaskSettingsPayload) {
    validateTaskPayload(payload);
    const task = this.getTask(taskId);
    const now = nowInAppTimezone();
    const nextRunAt = task.enabled ? getNextRunAt(payload.schedule, now).toISOString() : null;
    return this.taskRepository.updateTask(taskId, {
      schedule: payload.schedule.trim(),
      prompt: payload.prompt.trim(),
      environmentVariables: normalizeEnvironmentVariables(payload.environment_variables),
      updatedAt: now.toISOString(),
      nextRunAt
    });
  }

  private createTaskRecord(
    taskId: string,
    threadId: string,
    workspaceDirectory: string,
    environmentVariables: Record<string, string>,
    payload: TaskPayload
  ) {
    const now = nowInAppTimezone();
    const nextRunAt = getNextRunAt(payload.schedule, now);
    return this.taskRepository.createTask({
      id: taskId,
      schedule: payload.schedule.trim(),
      threadId,
      prompt: payload.prompt.trim(),
      workspaceDirectory,
      environmentVariables,
      enabled: true,
      createdAt: now.toISOString(),
      nextRunAt: nextRunAt.toISOString()
    });
  }

  setEnabled(taskId: string, payload: TaskEnabledPayload) {
    if (typeof payload?.enabled !== "boolean") {
      throw new AppError(422, "enabled must be a boolean.");
    }

    const task = this.getTask(taskId);
    const now = nowInAppTimezone();
    const nextRunAt = payload.enabled ? getNextRunAt(task.schedule, now).toISOString() : null;
    return this.taskRepository.setEnabled(taskId, payload.enabled, now.toISOString(), nextRunAt);
  }

  deleteTask(taskId: string): void {
    this.getTask(taskId);
    this.taskRepository.deleteTask(taskId);
  }

  listExecutionHistory(taskId: string | null = null) {
    if (taskId) {
      this.getTask(taskId);
    }
    return this.taskRepository.listExecutionHistory(taskId);
  }

  async getTaskMessages(taskId: string) {
    const task = this.getTask(taskId);

    if (!this.codexHistoryService) {
      throw new AppError(503, "Codex App Server is not configured.");
    }

    return {
      task_id: task.id,
      thread_id: task.thread_id,
      messages: await this.codexHistoryService.listMessages(task.thread_id)
    };
  }

  async sendChatMessage(taskId: string, payload: TaskChatPayload) {
    const task = this.getTask(taskId);
    const message = payload?.message?.trim();

    if (!message) {
      throw new AppError(422, "message is required.");
    }

    const result = await this.codexService.sendPrompt(task.thread_id, message, task.workspace_directory, task.environment_variables);

    if (!result.success) {
      throw new AppError(502, result.errorMessage ?? "Failed to send message to Codex.");
    }

    return {
      task_id: task.id,
      thread_id: task.thread_id,
      message,
      response_text: result.responseText
    };
  }
}

function validateTaskPayload(payload: TaskPayload | TaskSettingsPayload): void {
  if (typeof payload?.schedule !== "string" || !payload.schedule.trim()) {
    throw new AppError(422, "schedule is required.");
  }

  if (typeof payload?.prompt !== "string" || !payload.prompt.trim()) {
    throw new AppError(422, "prompt is required.");
  }

  try {
    validateSchedule(payload.schedule.trim());
  } catch (error) {
    throw new AppError(422, error instanceof Error ? error.message : "Invalid CRON schedule.");
  }

  if (payload.environment_variables !== undefined) {
    if (!payload.environment_variables || typeof payload.environment_variables !== "object" || Array.isArray(payload.environment_variables)) {
      throw new AppError(422, "environment_variables must be an object.");
    }

    for (const [key, value] of Object.entries(payload.environment_variables)) {
      if (!key.trim()) {
        throw new AppError(422, "environment_variables keys must be non-empty strings.");
      }

      if (typeof value !== "string") {
        throw new AppError(422, "environment_variables values must be strings.");
      }
    }
  }
}
